be1-go/channel/election/mod.go

```diff
@@ -304,141 +310,149 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
 	return nil
 }
 
-func (c *Channel) publishCastVote(msg message.Message) error {
-	voteMsg, err := c.getAndVerifyCastVoteMessage(msg)
+func (c *Channel) processCastVote(msg message.Message) error {
+
+	senderBuf, err := base64.URLEncoding.DecodeString(msg.Sender)
 	if err != nil {
-		return xerrors.Errorf("failed to get and verify vote message: %v", err)
+		return xerrors.Errorf("failed to decode sender key: %v", err)
 	}
 
-	//This should update any previously set vote if the message ids are the same
-	c.inbox.StoreMessage(msg)
-	for _, q := range voteMsg.Votes {
+	senderPoint := crypto.Suite.Point()
+	err = senderPoint.UnmarshalBinary(senderBuf)
+	if err != nil {
+		return answer.NewError(-4, "invalid sender public key")
+	}
 
-		qs, ok := c.questions[q.Question]
+	// verify sender is an attendee or the organizer
+	ok := c.attendees.IsPresent(msg.Sender) || c.hub.GetPubKeyOrg().Equal(senderPoint)
+	if !ok {
+		return answer.NewError(-4, "only attendees can cast a vote in an election")
+	}
 
-		if !ok {
-			return answer.NewErrorf(-4, "no Question with question ID %s exists", q.Question)
-		}
+	var castVote messagedata.VoteCastVote
 
-		// this is to handle the case when the organizer must handle multiple
-		// votes being cast at the same time
-		qs.validVotesMu.Lock()
-		earlierVote, ok := qs.validVotes[msg.Sender]
-
-		// if the sender didn't previously cast a vote or if the vote is no
-		// longer valid update it
-		if err := checkMethodProperties(qs.method, len(q.Vote)); err != nil {
-			return xerrors.Errorf("failed to validate voting method props: %w", err)
-		}
+	err = msg.UnmarshalData(&castVote)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal cast vote: %v", err)
+	}
 
-		if !ok {
-			qs.validVotes[msg.Sender] = validVote{
-				voteMsg.CreatedAt,
-				q.Vote,
-			}
-		} else {
-			changeVote(qs, earlierVote, msg.Sender, voteMsg.CreatedAt, q.Vote)
-		}
+	err = c.verifyMessageCastVote(castVote)
+	if err != nil {
+		return xerrors.Errorf("invalid election#cast_vote message: %v", err)
+	}
 
-		// other votes can now change the list of valid votes
-		qs.validVotesMu.Unlock()
+	err = updateVote(msg.MessageID, msg.Sender, castVote, &c.questions)
+	if err != nil {
+		xerrors.Errorf("failed to update vote: %v", err)
 	}
 
+	c.inbox.StoreMessage(msg)
+	c.broadcastToAllClients(msg)
+
 	return nil
 }
 
-func (c *Channel) getAndVerifyCastVoteMessage(msg message.Message) (messagedata.VoteCastVote, error) {
-	var voteMsg messagedata.VoteCastVote
+func (c *Channel) processElectionEnd(msg message.Message) error {
 
-	err := msg.UnmarshalData(&voteMsg)
+	sender := msg.Sender
+	senderBuf, err := base64.URLEncoding.DecodeString(sender)
 	if err != nil {
-		return voteMsg, xerrors.Errorf("failed to unmarshal cast vote: %v", err)
+		c.log.Error().Msgf("problem decoding sender public key: %v", err)
```

Comment:
is there a reason why we log this specific error and not always the others ?

Reply:
It was to debug at a moment, I can add logs to every error if it's best

Reply:
There should be a log only once at the top level, if any.